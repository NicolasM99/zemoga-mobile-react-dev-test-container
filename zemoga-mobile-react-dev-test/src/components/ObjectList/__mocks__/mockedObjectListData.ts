export const mockedObjectList = {
  test: 'Test attribute',
  nested0: {
    nestedlvl0attribute: 'Nested level 0 attribute',
    nested1: {
      nestedlvl1attribute: 'Nested level 1 attribute'
    }
  }
};
