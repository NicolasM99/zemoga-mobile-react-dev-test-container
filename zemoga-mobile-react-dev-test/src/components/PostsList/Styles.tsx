import { StyleSheet } from 'react-native';

export const postsListStyles = StyleSheet.create({
  postsListContainer: {
    flex: 1,
    width: '100%'
  }
});
