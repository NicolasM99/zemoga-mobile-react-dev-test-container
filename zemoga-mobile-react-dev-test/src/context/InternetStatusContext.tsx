import React, { createContext, FC, useContext, useEffect, useState } from 'react';

import NetInfo from '@react-native-community/netinfo';

import { IInternetStatusContext, InternetStatusContextType } from './@types/internetStatusContext';

const InternetStatusContext = createContext<InternetStatusContextType | null>(null);

const InternetStatusProvider: FC<IInternetStatusContext> = ({
  children
}: IInternetStatusContext) => {
  const [internetStatus, setInternetStatus] = useState<boolean | null>(true);
  useEffect(() => {
    const unsubscribe = NetInfo.addEventListener((state) => {
      setInternetStatus(state.isConnected);
    });
    return () => {
      unsubscribe();
    };
  }, []);

  const valueObj = {
    internetStatus
  };
  return (
    <InternetStatusContext.Provider value={valueObj}>{children}</InternetStatusContext.Provider>
  );
};

const useInternetStatusContext = () => useContext(InternetStatusContext);

export { InternetStatusContext, InternetStatusProvider, useInternetStatusContext };
