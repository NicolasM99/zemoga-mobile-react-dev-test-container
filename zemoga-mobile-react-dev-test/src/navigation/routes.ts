export const POSTS = 'Posts';
export const POST_DETAILS = 'PostDetails';
