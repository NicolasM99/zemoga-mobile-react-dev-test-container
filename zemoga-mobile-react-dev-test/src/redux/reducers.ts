import { IInitialState } from './@types/initialState';
import {
  CHANGE_VALUES,
  CLEAR_VALUES,
  GENERAL_ERROR,
  GET_COMMENTS,
  GET_POSTS,
  GET_USER
} from './actionTypes';

export const INITIAL_STATE: IInitialState = { posts: [], comments: [], user: {} };

export default (state: any, action: any) => {
  if (action.type === GENERAL_ERROR)
    return {
      ...state,
      status: action?.payload?.response?.status,
      students: action?.payload?.response?.data
    };
  if ([CHANGE_VALUES, CLEAR_VALUES].includes(action.type)) return { ...state, ...action.variables };
  if ([GET_POSTS].includes(action.type))
    return { ...state, status: action.payload.status, posts: action.payload.data };
  if ([GET_USER].includes(action.type))
    return { ...state, status: action.payload.status, user: action.payload.data };
  if ([GET_COMMENTS].includes(action.type))
    return { ...state, status: action.payload.status, comments: action.payload.data };
  //   if ([GET_STUDENTS].includes(action.type))
  //     return {
  //       ...state,
  //       status: action.payload.status,
  //       students: action.payload.data
  //     };
  //   if ([ADD_STUDENTS, DELETE_STUDENTS, UPDATE_STUDENT].includes(action.type))
  //     return {
  //       ...state,
  //       status: action.payload.status
  //     };
  return state;
};
