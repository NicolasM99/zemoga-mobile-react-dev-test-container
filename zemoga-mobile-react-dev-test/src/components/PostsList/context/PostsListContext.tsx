import React, { FC, useState, createContext, useContext, useEffect } from 'react';

import { generalAction } from 'src/redux/actions';
import { GET_POSTS } from 'src/redux/actionTypes';
import { useRedux } from 'src/redux/hooks/useRedux';
import { POSTS_API } from 'src/redux/paths';

import { PostsListProviderProps, PostsListContextType, IPost } from '../@types/postListContext';

const PostsListContext = createContext<PostsListContextType | null>(null);

const PostsListProvider: FC<PostsListProviderProps> = ({ children }: PostsListProviderProps) => {
  const [posts, setPosts] = useState<IPost[] | []>([]);
  const [postsToDelete, setPostsToDelete] = useState<IPost[] | []>([]);
  const [isDeletingItems, setIsDeletingItems] = useState(false);
  const [dispatch, { posts: postsStore }, status] = useRedux(['posts']);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    if (postsStore) setPosts([...postsStore]);
  }, [postsStore]);

  useEffect(() => {
    if (status) setIsLoading(false);
  }, [status]);

  useEffect(() => {
    if (isLoading) setIsLoading(false);
  }, [isLoading]);

  useEffect(() => {
    dispatch(generalAction({ actionType: GET_POSTS, api: POSTS_API }));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  const valueObj = {
    posts,
    setPosts,
    isDeletingItems,
    setIsDeletingItems,
    postsToDelete,
    setPostsToDelete,
    isLoading,
    setIsLoading
  };
  return <PostsListContext.Provider value={valueObj}>{children}</PostsListContext.Provider>;
};

const usePostsListContext = () => useContext(PostsListContext);

export { PostsListContext, PostsListProvider, usePostsListContext };
