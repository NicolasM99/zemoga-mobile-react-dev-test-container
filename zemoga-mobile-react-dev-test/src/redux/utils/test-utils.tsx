import React, { PropsWithChildren } from 'react';

import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { render } from '@testing-library/react-native';
import type { RenderOptions } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import { INITIAL_STATE } from '../reducers';
import { persistedReducer, store as reduxStore } from '../store';
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<typeof persistedReducer>;
  store?: typeof reduxStore;
}

export const renderWithStoreProviders = (
  ui: React.ReactElement,
  {
    preloadedState = INITIAL_STATE,
    store = configureStore({ reducer: persistedReducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren<object>): JSX.Element => (
    <Provider store={store}>{children}</Provider>
  );

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
