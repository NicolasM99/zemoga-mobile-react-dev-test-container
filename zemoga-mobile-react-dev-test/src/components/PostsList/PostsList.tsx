import React, { FC, useEffect, useState, useCallback } from 'react';

import { useNavigation } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { View, Alert, FlatList } from 'react-native';

import { InternetStatusContextType } from 'src/context/@types/internetStatusContext';
import { useInternetStatusContext } from 'src/context/InternetStatusContext';
import { POST_DETAILS } from 'src/navigation/routes';
import { StackParamList } from 'src/navigation/Stacks';
import { IPostStore } from 'src/redux/@types/postStore';
import { changeValues, generalAction } from 'src/redux/actions';
import { DELETE_POST, PATCH_POST } from 'src/redux/actionTypes';
import { useRedux } from 'src/redux/hooks/useRedux';
import { POSTS_API } from 'src/redux/paths';

import { IPost, PostsListContextType } from './@types/postListContext';
import DeletePostsBtn from './components/DeletePostsBtn/DeletePostsBtn';
import DeleteUnfavouritePostsBtn from './components/DeleteUnfavouritePostsBtn/DeleteUnfavouritePostsBtn';
import Post from './components/Post/Post';
import { usePostsListContext } from './context/PostsListContext';
import { postsListStyles } from './Styles';
import {
  handleFilterDeletedPosts,
  handleSelectPost,
  handleSetFavouritePost
} from './util/postListHandlers';

const PostsList: FC = () => {
  const {
    isDeletingItems,
    setIsDeletingItems,
    posts,
    setPosts,
    postsToDelete,
    setPostsToDelete,
    setIsLoading
  } = usePostsListContext() as PostsListContextType;
  const [isLoadingDelete, setIsLoadingDelete] = useState<boolean>(false);
  const { internetStatus } = useInternetStatusContext() as InternetStatusContextType;
  const [dispatch, { patchedPost, requestIndex }, status] = useRedux([
    'patchedPost',
    'requestIndex'
  ]);
  const [deleteData, setDeleteData] = useState<IPost[] | []>([]);
  const [favouritePostIndex, setFavouritePostIndex] = useState<number | null>(null);

  const navigation = useNavigation<NativeStackNavigationProp<StackParamList>>();
  useEffect(() => {
    if (isDeletingItems) setPostsToDelete([...posts].filter((post) => post.isSelected));
  }, [posts]);

  useEffect(() => {
    if (status === 200) {
      if ((favouritePostIndex || favouritePostIndex === 0) && patchedPost) {
        const newPosts = handleSetFavouritePost([...posts], favouritePostIndex);
        dispatch(changeValues({ posts: newPosts, patchedPost: null }));
        setPosts(newPosts);
        setFavouritePostIndex(null);
        setIsLoading(false);
      }
      if (requestIndex === deleteData?.length + 1) {
        const newPosts = handleFilterDeletedPosts(posts, deleteData);
        dispatch(changeValues({ posts: newPosts, requestIndex: null }));
        setPostsToDelete([]);
        setDeleteData([]);
        setPosts(newPosts);
        setIsLoadingDelete(false);
        setIsDeletingItems(false);
        alert('Posts deleted successfully!');
      }
    }
  }, [status, favouritePostIndex, patchedPost, requestIndex]);

  useEffect(() => {
    if (requestIndex && requestIndex < deleteData?.length + 1) {
      dispatch(changeValues({ status: null }));
      dispatch(
        generalAction({
          actionType: DELETE_POST,
          method: 'delete',
          api: `${POSTS_API}/${deleteData[requestIndex - 1].postId}`,
          requestIndex
        })
      );
    }
  }, [requestIndex]);

  useEffect(() => {
    if (deleteData.length) dispatch(changeValues({ requestIndex: 1 }));
  }, [deleteData]);

  useEffect(() => {
    setIsLoading(isLoadingDelete);
  }, [isLoadingDelete]);

  const handlePressPost = useCallback(
    (post: IPostStore, index: number, isSelected: boolean) => {
      if (!isDeletingItems) {
        navigation.navigate(POST_DETAILS, { ...post, postId: post.id });
        return;
      }
      const newPosts = handleSelectPost([...posts], index, isSelected);
      setPosts(newPosts);
    },
    [isDeletingItems, navigation, posts, setPosts]
  );

  const handleDeletePosts = (data: IPost[]) => {
    setIsLoadingDelete(true);
    setDeleteData(data);
  };

  const handleConfirmDeletePosts = (onlyUnfavouritePosts = false) => {
    const auxDeleteData: IPost[] = onlyUnfavouritePosts
      ? [...posts].filter(({ isFavourite }) => !isFavourite)
      : [...postsToDelete];

    Alert.alert(
      'Delete posts',
      `Are you sure you want to delete ${auxDeleteData.length} posts?\n\nThis action is NOT reversible and may take more time than excpected for a large amount of items.`,
      [
        {
          text: 'Cancel',
          onPress: () => {},
          style: 'destructive'
        },
        { text: 'OK', onPress: () => handleDeletePosts(auxDeleteData) }
      ]
    );
  };

  const handleOnSetFavourite = useCallback(
    (index: number, id: number) => {
      setFavouritePostIndex(index);
      setIsLoading(true);
      dispatch(
        generalAction({
          actionType: PATCH_POST,
          api: `${POSTS_API}/${id}`,
          method: 'patch',
          data: { isFavourite: true }
        })
      );
    },
    [dispatch, setIsLoading]
  );

  const renderPostItem = useCallback(
    ({ item: { id, title, isFavourite, isSelected, ...rest }, index }: any) => (
      <Post
        title={title}
        key={id}
        onPressPost={() => handlePressPost({ id, title, ...rest }, index, isSelected)}
        isFavourite={isFavourite}
        onSetFavourite={() => handleOnSetFavourite(index, id)}
        isSelected={isSelected}
        isDeletingItems={isDeletingItems}
        actionsDisabled={!internetStatus}
      />
    ),
    [isDeletingItems, handleOnSetFavourite, handlePressPost, internetStatus]
  );

  return (
    <>
      <FlatList
        style={postsListStyles.postsListContainer}
        data={posts}
        renderItem={renderPostItem}
        initialNumToRender={10}
        maxToRenderPerBatch={5}
        removeClippedSubviews
        windowSize={5}
      />
      {isDeletingItems && (
        <View style={{ width: '100%', bottom: 0, position: 'absolute' }}>
          {postsToDelete.length > 0 && (
            <DeletePostsBtn onPress={() => handleConfirmDeletePosts()} />
          )}
          <DeleteUnfavouritePostsBtn onPress={() => handleConfirmDeletePosts(true)} />
        </View>
      )}
    </>
  );
};

export default PostsList;
