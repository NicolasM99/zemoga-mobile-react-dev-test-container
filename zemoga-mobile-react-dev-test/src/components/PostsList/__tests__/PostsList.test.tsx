import { fireEvent, waitFor, screen } from '@testing-library/react-native';

import * as InternetStatusContext from 'src/context/InternetStatusContext';
import { POST_DETAILS } from 'src/navigation/routes';
import { INITIAL_STATE } from 'src/redux/reducers';
import { renderWithStoreProviders } from 'src/redux/utils/test-utils';
import MockedScreenProvider from 'src/screens/__mocks__/MockedScreenProvider';

import { mockedPosts } from '../__mocks__/mockedData';
import PostsList from '../PostsList';

const mockedNavigate = { navigate: () => true };
jest.mock('@react-navigation/native', () => ({
  ...(jest.requireActual('@react-navigation/native') as any),
  useNavigation: () => mockedNavigate
}));
jest.mock('expo-file-system', () => ({
  ...(jest.requireActual('expo-file-system') as any),
  getInfoAsync: async () => Promise.resolve(true)
}));
describe('<PostsList /> unit tests', () => {
  jest.mock('react-redux', () =>
    jest.fn(() => ({
      useDispatch: () => true
    }))
  );
  jest.spyOn(InternetStatusContext, 'useInternetStatusContext').mockReturnValue({
    internetStatus: true
  });
  it('Should mount the component correctly', async () => {
    renderWithStoreProviders(
      <MockedScreenProvider>
        <PostsList />
      </MockedScreenProvider>
    );
    expect(screen).toMatchSnapshot();
  });
  it('Should render a list of 10 posts correctly', async () => {
    const { findAllByTestId } = renderWithStoreProviders(
      <MockedScreenProvider>
        <PostsList />
      </MockedScreenProvider>,
      { preloadedState: { ...INITIAL_STATE, posts: mockedPosts } }
    );
    const postTitles = await findAllByTestId('postTitle');
    postTitles.map((post, index) => {
      waitFor(() => expect(post.props.children).toEqual(mockedPosts[index].title));
    });
  });

  it('Should go to PostDetails screen with correct data when clicking a post', async () => {
    spyOn(mockedNavigate, 'navigate');
    renderWithStoreProviders(
      <MockedScreenProvider>
        <PostsList />
      </MockedScreenProvider>,
      { preloadedState: { ...INITIAL_STATE, posts: mockedPosts } }
    );
    const postItems = await screen.findAllByTestId('postItem');
    const postItem = postItems[0];
    fireEvent.press(postItem);
    expect(mockedNavigate.navigate).toHaveBeenCalledWith(POST_DETAILS, {
      ...mockedPosts[0],
      postId: mockedPosts[0].id
    });
  });
});
