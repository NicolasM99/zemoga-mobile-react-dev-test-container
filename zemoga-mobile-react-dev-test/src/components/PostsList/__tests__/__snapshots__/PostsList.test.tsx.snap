// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostsList /> unit tests Should mount the component correctly 1`] = `
<RCTScrollView
  data={Array []}
  getItem={[Function]}
  getItemCount={[Function]}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={5}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={true}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
  viewabilityConfigCallbackPairs={Array []}
  windowSize={5}
>
  <View />
</RCTScrollView>
`;
